---
title: "Introduction to the MethSig pacakge"
author: "Heng Pan"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup, echo = F}
library(MethSig)
```

## Setup
```{r, eval = F}
setRepositories(ind = 1:2)
install.packages("devtools")
devtools::install_github('HengPan2007/MethSig', build_vignettes = T)
library(MethSig)
```

## Introduction
MethSig is used to infer DNA methylation drivers from bisulfite converted data of cancer cohorts.

## Prerequisite

### 1. Data alignment and processing.
Bisulfite converted data are aligned and processed as described in the published book chapter (Pan *et al.*, *Cancer Systems Biology*, 2018).

### 2. Covariates matrix Preparation.
A matrix of covariates, which are used in the beta regression model, is needed. The matrix needs to include a column of hugo gene symbol and at least one column of the covariate. An example can be loaded by *invisible(matCV)*.

```{r, eval = F}
matCV <- invisible(matCV)
```
This example matrix includes Hugo symbol (Hugo), average promoter DHcR level in control samples (DHcR_Normal), average promoter PDR level in control samples (PDR_Normal), average gene expression level in control samples (GEXP_Normal) and DNA replication time (Reptime). Users can define their own covariates.

```{r, echo = F}
head(matCV)
```

### 3. Files containing information of hypermethylated cytosine (HCs).
A tab-separated values input file (without a header line). The input file contains details of differentially methylated cytosines (DMCs) with following columns (V1 to V11): chr, pos, numC in control, numC + numT in control, numC in tumor, numC + numT in tumor, CpG methylation ratio (tumor methylation / control methylation), chi-squared test p-value, adjusted p-value, significance, hyper or hypo in tumor. Details of generating this type of files were described in Pan *et al.*, *Cancer Systems Biology*, 2018. An example file can be found in extdata/DMC.SRR2069925.txt. Notably, the name needs to be in DMC.Sample_name.txt format.

```{r, echo = F}
tmp <- read.table(system.file("extdata", "DMC.SRR2069925.txt", package = "MethSig"), sep='\t')
head(tmp)
```

### 4. Files containing information of proportion of discordant reads (PDR) at CpGs 
A tab-separated values input file. The input file contains details of single CpG PDR with following columns: chr, start, strand, ConMethReadCount, ConUMethReadCount, DisReadCount, NAReadCount. Details of PDR were described in Landau *et al.*, *Cancer Cell*, 2014. extdata/pdrCall_from_Bismark.py can be used to call PDR of single CpG from Bismark (Krueger *et al.*, *Bioinformatics*, 2011) outputs files starting with CpG_OB or CpG_OT. An example file can be found in extdata/PDR.SRR2069925.txt. Notably, the name needs to be in PDR.sample_name.txt format.

```{r, echo=F}
tmp <- read.table(system.file("extdata", "PDR.SRR2069925.txt", package = "MethSig"), header=T, sep='\t')
head(tmp)
```

## Make input matrix

### 1. Calculate promoter DHcR
Promoter (defined as Â± 2kb windows centered on RefSeq transcription start site) hypermethylation is measured using differentially hypermethylated cytosine ratio (DHcR), defined as the ratio of hypermethylated cytosines (HCs) to the total number of promoter CpGs profiled. HCs of each sample are defined as CpGs at which DNAme is statistically higher than the average DNAme of control samples (false discovery rate = 20%, Chi-squared test). Only CpGs with read depth greater than 10 are included in the analysis. RRBS data of normal samples are used as control. An implemented function *makeHG19Promoters* can be used to provide promoter annotations of hg19 RefSeq genes. Users and define their own annotation.

```{r}
dhcr <- promoterDHcR(file_name = system.file("extdata", "DMC.SRR2069925.txt", package = "MethSig"),
                     pro = makeHG19Promoters())
head(dhcr)
```

### 2. Calculate promoter PDR
If all the CpGs on a specific read are methylated, or all the CpGs on a read are unmethylated, the read is classified as concordant; otherwise it is classified as discordant. At each CpG, the PDR is equal to the number of discordant reads that cover that location divided by the total number of reads that cover that location. The PDR of promoter is given by averaging the values of individual CpGs, as calculated for all CpGs within the promoter of interest with read depth greater than 10 reads and that are covered by reads that contain at least 4 CpGs.

```{r}
pdr <- promoterPDR(file_name = system.file("extdata", "PDR.SRR2069925.txt", package = "MethSig"),
                   pro = makeHG19Promoters())
head(pdr)
```

### 3. Make input matrix
As mentioned in **Prerequisite** section, users need to put DMC.sample_name.txt and PDR.sample_name.txt files in the input_dir folder. Also, a user defined covarites matrix is needed.
```{r, eval = F}
ds <- makeInputMatrix(names_list = as.list('SRR2069925'),
                matCV = invisible(matCV),
                pro = makeHG19Promoters(),
                input_dir = system.file("extdata", "", package = "MethSig"))
head(ds)
```
```{r, echo = F}
head(invisible(inputMat))
```
### Sample-specific hypermethylation inference
Expected promoter DHcR of tumor samples is estimated by beta regression model and expected DHcR is tested against observed DHcR to infer hypermethylation status.
```{r, eval = F}
pval <- pvalueBetaReg(formula = as.formula('DHcR_Tumor_Beta~DHcR_Normal+PDR_Normal+GEXP_Normal+Reptime+PDR_Tumor+Depth_Tumor+CpGs_Tumor'),
                      data = invisible(inputMat))
head(pval)
```
```{r, echo = F}
head(invisible(pvalByGenePt))
```

### Cohort-prevalent hypermethylation inference
Wilkinson p-value combination method is used to determine if promoter hypermethylation is over-represented in the cohort. To eliminate the effect of cohort size on p-value combination results, MethSig randomly samples equal number of patients iteratively and uses lower quartile of combined p-values to infer hypermethylation.
```{r}
pval <- pvalueCombine(data = invisible(pvalByGenePt))
head(pval)
```
